drivers/adc_ng/adc_ng.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (C) 2020 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     drivers_adc_ng
+ * @{
+ *
+ * @file
+ * @brief       Implementation of common functions of the ADC API
+ *
+ * @author      Marian Buschsieweke <marian.buschsieweke@ovgu.de>
+ * @}
+ */
+
+#include <errno.h>
+
+#include "adc_ng.h"
+
+#if (ADC_NG_NUMOF == 1) && defined(MODULE_PERIPH_ADC_NG)
+extern const adc_ng_driver_t periph_adc_ng_driver;
+const adc_ng_backend_t adc_ng_backends[ADC_NG_NUMOF] = {
+    {
+        .driver = &periph_adc_ng_driver,
+        .handle = NULL
+    }
+};
+#endif
+/* selected reference voltage in mV */
+int16_t adc_ng_refs[ADC_NG_NUMOF];
+/* selected resolution */
+uint8_t adc_ng_res[ADC_NG_NUMOF];
+
+int adc_ng_init(uint8_t adc, uint8_t chan, uint8_t res, int16_t *ref)
```

Comment:
I wonder how much overhead it would be to just have

```suggestion
int adc_ng_init(gpio_t pin, uint8_t res, int16_t *ref)
```

and not have the user deal with implementation details of the GPIO.

Reply:
Note that the ADC can be connected to internal sources, such as as NTC for temperature measurement. Also, there are differential channels that are connected to two inputs. (The sample will be positive if the "plus" side of the channel has a higher voltage than the "minus" size, and a negative sample if the "minus" side is higher.)

By having pins as entities, only a subset of the features of an ADC can be provided. (See the lengthy discussion above fore more details.)

Reply:
Also not that external ADCs would no longer be possible with that, while the current API is fully compatible.